src/request_queue.js

```diff
@@ -240,165 +268,225 @@ export class RequestQueue {
         const cachedInfo = this.requestsCache.get(cacheKey);
 
         if (cachedInfo) {
-            return Promise.resolve({
+            return {
                 wasAlreadyPresent: true,
                 // We may assume that if request is in local cache then also the information if the
                 // request was already handled is there because just one client should be using one queue.
                 wasAlreadyHandled: cachedInfo.isHandled,
                 requestId: cachedInfo.id,
                 // TODO: Why not set request.id to cachedInfo.id???
                 request,
-            });
+            };
         }
 
-        return requestQueues
-            .addRequest({
-                request,
-                queueId: this.queueId,
-                forefront,
-                clientKey: this.clientKey,
-            })
-            .then((queueOperationInfo) => {
-                const { requestId, wasAlreadyHandled, wasAlreadyPresent } = queueOperationInfo;
+        const queueOperationInfo = await requestQueues.addRequest({
+            request,
+            queueId: this.queueId,
+            forefront,
+            clientKey: this.clientKey,
+        });
 
-                this._cacheRequest(cacheKey, queueOperationInfo);
+        const { requestId, wasAlreadyPresent } = queueOperationInfo;
 
-                if (forefront && !this.requestIdsInProgress[requestId] && !wasAlreadyHandled) {
-                    this.queueHeadDict.add(requestId, requestId, true);
-                }
+        this._cacheRequest(cacheKey, queueOperationInfo);
 
-                if (!wasAlreadyPresent) {
-                    this.assumedTotalCount++;
-                }
+        if (!wasAlreadyPresent && !this.inProgress[requestId] && !this.recentlyHandled.get(requestId)) {
+            this.assumedTotalCount++;
 
-                // TODO: Why not set request.id to cachedInfo.id???
-                queueOperationInfo.request = request;
+            // Performance optimization: add request straight to head if possible
+            if (forefront) {
+                this.queueHeadDict.add(requestId, requestId, true);
+            } else if (this.assumedTotalCount < QUERY_HEAD_MIN_LENGTH) {
+                this.queueHeadDict.add(requestId, requestId, false);
+            }
+        }
 
-                return queueOperationInfo;
-            });
+        // TODO: Why not set request.id to cachedInfo.id???
```

Comment:
Could resolve this TODO. Tbh I no longer know who added it and what it means.

Reply:
I did, but you're right, we should fix this too.